### DISCUSSION TITLE
qtranslate slug

### DISCUSSION THREAD

**Discussion by andreysneg** [qtranslate/qtranslate-xt#1106] [Q&A] [answered]
Hello. 

Qtranslate cannot be used without "Qtranslate slug" plugin. All SEO die without slugs for every language.

Now "Qtranslate slug" plugin removed from wordpress repository.
And last update was 6 years ago.

Github of project here. But github plugin have many errors and dont work normally.
https://github.com/not-only-code/qtranslate-slug

Here my slightly modified for php 8.1 plugin version:
https://sneg.ee/qtranslate-slug-php8.1.zip

I suggest either inserting this plugin directly into the Qtranslate (with the permission of the developer) or simply rewriting it with an alternative code. This is small plugin. 

Thank you.

**Comment by herrvigg** [comment]
We opened this discussion long ago, see #671.
It is happening right now. A fork of `qtranslate-slug` has just been merged as an experimental module, see #1060. The integration is still in progress, but please test it.

**Comment by Hr0bar** [comment]
Hi, I see its been integrated in the master branch and soon to be released as part of qTranslateXT.

Is there any special prodecure in how to uninstall "qTranslate slug" before upgrading to latest qTranslateXT which includes it now?

Thanks

**Reply by herrvigg** [reply]
Yes it will be released soon!
I don't know what the uninstall of slugs does, but QTX will import the slugs data on activation. To enable the new slugs module it will be necessary to disable the slugs plugin, but QTX will notify you about that.

@spleen1981 In fact there's one thing we didn't think of, when slugs is uninstalled it would be problematic if the metadata (related to posts and tags) are erased because we still share the same keys. The settings are already cloned in separate option keys, but not the metadata . Should we do something special here?

**Reply by spleen1981** [reply]
> Is there any special prodecure in how to uninstall "qTranslate slug" before upgrading to latest qTranslateXT which includes it now?

Good point, if the legacy QTS plugin is uninstalled from WP plugins page, posts metadata will be [definitely removed](https://github.com/not-only-code/qtranslate-slug/blob/386b90723d79792ffe4b88138908a065e82b01f2/qtranslate-slug.php#L178-L182).
Hence we should mention in migration notes that legacy plugin folder needs to be deleted manually rather then uninstalled from WP admin, this is the safest and easiest solution...
And a backup of the database before is strongly suggested as always, of course...

No issue if legacy plugin is only deactivated.

**Reply by herrvigg** [reply]
Now that we move away from the plugin I believe we need a better solution. We should not rely or be dependent on the data owned by the plugin. Imo we need an import step that would duplicate all in our own QTX keys. Once the import is done, the plugin can be uninstalled and its own data will be erased.

**Reply by herrvigg** [reply]
The question is how should the import be done? At QTX plugin activation? With a manual button in the slugs settings? Or somewhere else?

**Reply by Hr0bar** [reply]
As a common plugin user, it would be ideal to have this happen automatically/on forced prompt question when its detected that "qTranslate slug" is installed on qtranslate-xt activation AND automatically/on forced prompt question on update of qtranslate-xt version that adds the slug integration, so next version to be released (but the on forced prompt may not be technically possible im afraid in wordpress - what if admin does not visit the page etc, only replaces plugin folder and so on)

So I vote for **"manual button in the slugs settings"**. Also gives more control to the admins instead of doing automagical things automatically, which many people do not like. Also if the admin is removing the old plugin then he is already consciously tinkering with the wordpress instance, so it makes sense to be more manual on QTS side as well then.

**Reply by spleen1981** [reply]
If we want to provide the migration feature from qtx settings, a button to remove the old plugin folder included in a notice on slugs module activation could be enough...
I would avoid messing up with existing metadata from posts terms etc.
Even if we do, that should be triggered by a manual action given the implications (no way back).

**Reply by herrvigg** [reply]
QTX should never uninstall any other plugin. This action is entirely left to the admin.
What do you mean by messing up metadata? The current state is messed up because there's no clear ownership. The "QTS" data are now owned by both the QTS and QTX plugins. QTX is a different plugin, it's too risky to say "don't uninstall QTS otherwise you lose all in QTX as well".

It's technically feasible to have a prompt question, we have such feature already in QTX. The admin button could also be interesting for more advanced admins, but I don't think it should be in the main slugs settings because it's not something you should act on normally.

**Reply by herrvigg** [reply]
Compared to the other modules it's a bit unfair for QTS because it has a clean uninstall that erases its own data. For the other legacy plugins such as ACF, WooCommerce, ... there's nothing similar for what I know. If QTS didn't have a proper uninstall it wouldn't be a problem 😓

**Reply by herrvigg** [reply]
Resolved by #1171 migration of QTS.

**Comment by spleen1981** [comment]
> What do you mean by messing up metadata? 

Just that it would be safer to keep current metadata. Also one could go back if needed for any reason.

> QTX should never uninstall any other plugin.

Agreed in principle, however here we are talking about a legacy external module now merged in QTX, so no big deal IMO.

Anyway if we want to avoid the old plugin removal from QTX and keep the thing simple (no modifications to database), we can just remove the action at plugin uninstallation, like done in #1167.
In this way user can just uninstall the old plugin with no harm and without having to do anything special.

**Reply by herrvigg** [reply]
See new QTS import feature in #1171.

**Comment by herrvigg** [comment]
The import feature from QTS has become a migration with #1171, now merged in master. The release 3.12.0 is approaching. Please test this and let us know how it goes. 

There's a temporary hack to revert the renaming of `qts_options` in master before migration. That's meaningful for those who worked with the `slugs` module on master recently cc @bagaweb. The migration is expected to be done once from this point. This hack will be removed before release so those not using master in development are not impacted by this.

**Reply by herrvigg** [reply]
For precision: the hack to rename `qts_options` is just to ensure the current post type / taxonomies slugs are not lost during the migration (user data in the slugs settings). It doesn't affect the `slugs` functionality after migration.

**Comment by bagaweb** [comment]
Thanks for all your work!
I'm going to test the new master and check if everything works fine. Will test it both on the website where I was testing last week, and on another website which still uses the old slugs plugin.

**Comment by bagaweb** [comment]
I don't understand if the second checkbox should be checked or not...

_Confirm migration of QTS slugs in database. Attention! Existing slugs are replaced. Leave unchecked for a dry-run._

**Reply by spleen1981** [reply]
The second checkbox it's a confirmation. Check it if you really want to perform the migration (it's not reversible).
If you want just see the number of item that can be migrated, leave the second checkbox unchecked.

**Reply by herrvigg** [reply]
@bagaweb I thought it was clear enough. How would you write it otherwise? 

Is this better? _Confirm migration of QTS slugs. Attention! Existing slugs are replaced. Leave unchecked for a dry run test without change in database._

**Reply by herrvigg** [reply]
Maybe with a notice for backup.

_Confirm migration of QTS slugs. Attention! Existing slugs are replaced. Leave unchecked for a dry run test. Before confirmation, it is recommended to backup your database._

**Reply by herrvigg** [reply]
"Existing slugs are replaced." may sound confusing. It's quite complicated to summarize it all briefly. Perhaps it's better remove it and keep just this:

_Confirm to perform the migration of QTS slugs. Attention! It is recommended to backup your database before confirmation. Leave unchecked for a dry run test to see how many items are erased / migrated._

**Reply by spleen1981** [reply]
I think we could remove all mentions to "erased/deleted items" in the notice to make it simpler, and just reporting the migrated ones.
Adding to the second checkbox the disclaimer "All current module slugs data (if any) will be replaced." after the first sentence above should be sufficient.

**Reply by herrvigg** [reply]
In reality it's not all current data that are replaced, but only those that would conflict with the migrated ones.

**Reply by herrvigg** [reply]
Updated as:
_Confirm to perform the migration of QTS slugs, or leave unchecked for a dry run test without update. Attention! Current module slugs data may be overwritten._

**Comment by bagaweb** [comment]
Alright thanks for clarifying this.

I've tried the migration leaving the second checbox unchecked and this is the result:

`Dry-run mode:
Deleted 0 slugs from wp_postmeta (qtranslate_slug_).
Migrated 52 slugs from wp_postmeta (_qts_slug_).
Deleted 0 slugs from wp_termmeta (qtranslate_slug_).
Migrated 4 slugs from wp_termmeta (_qts_slug_).
Deleted 2 types from options.
Migrated 9 types from options.`

The slugs seems to be gone, should I run the migration again checking the second checkbox?

**Reply by spleen1981** [reply]
All the `Deleted` are the items created with the new module already existing before the import (if you don't have any it's ok).
All the `Migrated` are the old items to be imported. Check the second checkbox and save to proceed.
Better doing a db backup as a general rule before this kind of operations anyways..

**Reply by herrvigg** [reply]
"Deleted 2 new types from options" come from the last options saved at the same time the migration is requested. This happens before the migration is executed. It is a bit surprising you don't have the same number. This could be because you disabled other plugins during the migration (e.g. WooCommerce). It doesn't impact the migration though.

**Reply by bagaweb** [reply]
Thanks for your support.

I went through a second migration checking both checkboxes. 
Everything seemed to be working fine, then I found out there is still an issue, subpages are not working:

mywebsite.com/pagina/sottopagina/ => works fine
mywebsite.com/en/page/subpage/ => 404 error

Tried to rewrite permalinks and to update subpages but no success.

**Reply by herrvigg** [reply]
That should not be related to the migration, but there might be some regression in the `slugs` module vs QTS plugin. Is this happening in the QTS plugin as well?

**Reply by bagaweb** [reply]
It was working fine with the old QTSlug plugin.

**Reply by spleen1981** [reply]
Just tried subpages slugs with `site/genitore/figlio` and `site/en/parent/child` and seems to work on my config.
Slugs for that language must be defined both for parent and child, try to update both pages.
Maybe you can double check and open a separate issue if needed.

**Reply by bagaweb** [reply]
I tried to update both pages but still 404 on child page. Parent page works fine.

**Reply by spleen1981** [reply]
Can't reproduce, in my test subpages are working properly, query vars passed are like `pagename="base_page%2Fsub_page"`, where `base_page` and `sub_page` are both translated slugs.
Please open an issue if problem persists (providing as much details as possible, including actual url tested and configuration details from `QTX settings > Troubleshooting > Collect information`), this discussion does not belong here...

**Reply by bagaweb** [reply]
Just to make things clear before opening a new thread... Are you coming from an old qTranslate & qTranslate slug or have you started from scratch from a fresh install?

**Reply by bagaweb** [reply]
Just tried on another website on another server, upgrading from qTranslate-X 3.4.6.8 & qTranslate slug 1.1.18, after migration all child pages give 404.

I rolled back the database, deactivated the module "Slugs translation (experimental)" of qTranslate‑XT, reinstalled and reactivated qTranslate slug 1.1.18, everything works ok.

**Reply by spleen1981** [reply]
Existing test environment with old slugs plugin + qtx latest master -> migrated to slugs module -> tested as above.
I've added a new subpage, linked to an existing page.

**Reply by bagaweb** [reply]
Upgrading from qTranslate-X 3.4.6.8 might be causing the issue then.

**Reply by bagaweb** [reply]
I'm gonna try another enviroment: qTranslate-XT 3.10.0 & qTranslate slug 1.1.18, upgrading to qtx latest master and let you know in a while.

**Reply by bagaweb** [reply]
I confirm that also the latest environment failed.

**Reply by herrvigg** [reply]
If this works with qtranslate-slug then there's a regression with the `slugs` module.  The problem is certainly not the migration that is just renaming data, but one change in `master` before this, it could have happened several months ago.

Are you familiar with `git`? If you go back to a DB before migration and pick the very first patch with `slugs` it would be interesting to see what happens.

I would try this:
- Get the very first QTS patch: `git checkout 9d1a8d831625554234efc09f9042eab8201bf773`. It's named "Integrate qtranslate-slug as experimental module" from Feb 14, 2022.
- If this works (it should), take a more recent commit SHA from that point in the [git history](https://github.com/qtranslate/qtranslate-xt/commits/master?after=69a9fa0ae541ca0e20b09f204025b359d024d942+69&branch=master&qualified_name=refs%2Fheads%2Fmaster), test the page.
- move forward in time towards more recent patches and repeat the test after each checkout.

Once you have narrowed the problematic commit it should be possible to understand the origin of the regression.

**Reply by bagaweb** [reply]
Thanks for your help, unfortunately I'm not familiar with git... 

Anyway that's what I tried:
- rolled back the DB;
- downloaded this zip https://github.com/qtranslate/qtranslate-xt/archive/9d1a8d831625554234efc09f9042eab8201bf773.zip;
- this one works as you said;
- tried all commits up to ebea260d341fb5ed3457b692a069aab4f1ce5828;
- all of them work.

Where should I move next?

**Reply by spleen1981** [reply]
you can go to newer commits, bottom of the page.

**Reply by bagaweb** [reply]
Spent a few hours trying all commits until latest one. Found out latest master works fine until I don't proceed with migration. So I suppose the issue is in the migration itself.

**Reply by herrvigg** [reply]
Oh really... that was surprising... Doesn't make sense to me 😕 
@spleen1981 is this related to meta or options? Meta I guess?

**Reply by herrvigg** [reply]
Has this something to do with hierarchical posts specifically?

**Reply by herrvigg** [reply]
@bagaweb if you go to edit the page in admin mode for the one that is failing, what slugs do you have? Then for the parent? Can you check all is correct there from the admin side?
> mywebsite.com/pagina/sottopagina/ => works fine
> mywebsite.com/en/page/subpage/ => 404 error

**Reply by spleen1981** [reply]
@herrvigg yes it's post meta.
Based on the feedbacks I tried to add a subpage before migration (my understanding is that issue is specific to hierarchical pages).
After migration everything is still working properly with my config.

**Reply by herrvigg** [reply]
Maybe something wrong in `get_page_uri`? What is confusing me is that I don't see how this could be related to that migration patch, for the post/page meta it should all be the same after a simple rename.

@bagaweb for the page that is failing can you retrieve the ID, run this in SQL and post the results?
```
SELECT * FROM wp_postmeta WHERE post_id=<ID>
```
There should be a unique slug by language, all keys prefixed with `qtranslate_`. There should not be any `_qts_` prefix left in that table.

**Reply by bagaweb** [reply]
> Has this something to do with hierarchical posts specifically?

Child pages

> @bagaweb if you go to edit the page in admin mode for the one that is failing, what slugs do you have? Then for the parent? Can you check all is correct there from the admin side?

All child pages are failing, not just one. All slugs are fine from the admin side. All slugs work fine on front end if I don't proceed with the migration.



> @bagaweb for the page that is failing can you retrieve the ID, run this in SQL and post the results?
> 
> ```
> SELECT * FROM wp_postmeta WHERE post_id=<ID>
> ```
> 
> There should be a unique slug by language, all keys prefixed with `qtranslate_`. There should not be any `_qts_` prefix left in that table.

I ran this in SQL before migrating and this is the result:

**meta_key:**
_qts_slug_it
_qts_slug_de
_qts_slug_en

**meta_value:**
_cantina-vini
weinkeller
wine-cellar_

Should I run it again after migrating?

**Reply by herrvigg** [reply]
> cantina-vini

Ottima scelta ;) 

> _qts_slug_it
_qts_slug_de
_qts_slug_en

This looks all good. After migration you should have :
qtranslate_slug_it
qtranslate_slug_de
qtranslate_slug_en

I don't think the problem comes from here.
Can you also check that the slugs appear correctly in the admin metabox, before and after migration?

**Reply by bagaweb** [reply]
> > cantina-vini
> 
> Ottima scelta ;)


Good wines indeed!


> I don't think the problem comes from here. Can you also check that the slugs appear correctly in the admin metabox, before and after migration?

Just found out that in the admin metaboxes the slugs are empty (using the latest master).

Tried again with 9d1a8d831625554234efc09f9042eab8201bf773 and the slugs are in place (in the admin metaboxes).

**Reply by herrvigg** [reply]
Mmm, at least this is consistent, if you have empty slugs in admin there's definitely something wrong here.
If you take commit 3a54f9e07c30b1d339ede47bcc076215374329dc just before migration, can you confirm it still works?
After migration do you have the expected meta with `qtranslate_slug` prefix?

**Reply by herrvigg** [reply]
Do you have other plugins that could make use of the QTS plugin? Own code?

**Reply by bagaweb** [reply]
> If you take commit [3a54f9e](https://github.com/qtranslate/qtranslate-xt/commit/3a54f9e07c30b1d339ede47bcc076215374329dc) just before migration, can you confirm it still works? 

I confirm it is working with this commit, slugs are ok both in front- and back-end.

Cannot see the migration checkboxes though.

> Do you have other plugins that could make use of the QTS plugin? Own code?

No

**Reply by herrvigg** [reply]
Great that we found the commit creating the regression in your case. This helps to nail it down.
I see there are some special cases when the post (page) status is in `draft`, `pending`, `auto-draft`. Is your page published?
After migration, if you run the SQL query on parent and child, do you have the expected prefix = `qtranslate_slug_`?
Could there be some cache somewhere? Be sure you don't have other plugins enabled so we eliminate external problems.

**Reply by herrvigg** [reply]
I made some tests on my side, but I have problems already before migration. To make it easier to track I opened #1180.

**Reply by herrvigg** [reply]
@bagaweb please can you create a separate issue for the case after migration? It will be easier to follow than in this "discussion" thread. It's also for better visibility for others. At first sight it looks different, but it _may_ have the same origin.

**Reply by bagaweb** [reply]
> Is your page published? 
Yes pages are published.

> @bagaweb please can you create a separate issue for the case after migration?
Is it ok if I reply to #1180 or should I open a new issue?

**Reply by herrvigg** [reply]
If you think it's a different issue, better open a new one. The problem mentioned in #1180 is about rewrite rules.

Please let us know what you have after migration for the child post here:
```
SELECT * FROM wp_postmeta WHERE post_id=<ID>
```

**Reply by bagaweb** [reply]
I think there's no need to open a new issue because the problem looks solved!

After migration I have :
qtranslate_slug_it
qtranslate_slug_de
qtranslate_slug_en

**Reply by herrvigg** [reply]
That looks correct!
How is the problem solved? With the last patch on master (from #1180)? In that case you can also comment in that issue, yes.

**Comment by herrvigg** [comment, marked as answer]
New module Slugs has been released in 3.12.0.
